inluxclient/query.go

```diff
@@ -0,0 +1,315 @@
+// Any remaining data in the current table is discarded.
+// When there are no more tables, it returns false.
+func (r *QueryResults) NextTable() bool {
+	return r.next(parsingModeTable)
+}
+
+// NextRow advances to the next row in the current table.
+// If called in the beginning, it also advances to the first table.
+// When there are no more rows in the current table, it returns false.
+func (r *QueryResults) NextRow() bool {
+	return r.next(parsingModeRow)
+}
+
+// Columns returns information on the columns in the current
+// table. It returns nil if there is no current table (for example
+// before NextTable has been called, or after NextTable returns false).
+func (r *QueryResults) Columns() []TableColumn {
+	return r.columns
+}
+
+// Err returns any error encountered. This should be called after NextTable or NextRow
+// returns false to check that all the results were correctly received.
+func (r *QueryResults) Err() error {
+	return r.err
+}
+
+// Values returns the values in the current row.
+// It returns nil if there is no current row.
+// All rows in a table have the same number of values.
+// The caller should not use the slice after NextRow
+// has been called again, because it's re-used.
+func (r *QueryResults) Values() []interface{} {
+	return r.values
+}
+
+// ValueByName returns value for given column name.
+// It returns nil if table has no value for such column.
+func (r *QueryResults) ValueByName(name string) interface{} {
+	if i, ok := r.names[name]; ok {
+		return r.values[i]
+	}
+	return nil
+}
+
+// next advances to next row/table via parsing csv rows.
+// It returns false at the end of each table  or at the end of the stream.
+func (r *QueryResults) next(mode parsingMode) bool {
+	// set closing query in case of preliminary return
+	var row []string
+	closer := func() {
+		r.columns = nil
+		r.values = nil
+		r.lastRow = nil
+		r.names = map[string]int{}
+		if err := r.Close(); err != nil {
+			message := err.Error()
+			if r.err != nil {
+				message = fmt.Sprintf("%s,%s", message, r.err.Error())
+			}
+			r.err = errors.New(message)
+		}
+	}
+	defer func() { closer() }()
+	parsingState := parsingStateDataRow
+	dataTypeAnnotationFound := false
+readRow:
+	if r.lastRow != nil {
+		row = r.lastRow
+		r.lastRow = nil
+	} else {
+		row, r.err = r.csvReader.Read()
+	}
+	if r.err == io.EOF {
+		r.err = nil
+		return false
+	}
+	if r.err != nil {
+		return false
+	}
+
+	if len(row) <= 1 {
+		goto readRow
+	}
+	if r.columns == nil {
+		parsingState = parsingStateNoTable
+	} else {
+		// test columns consistency in case of data row or already discovered annotations
+		if (row[0] == "" || parsingState == parsingStateAnnotation) && len(row)-1 != len(r.columns) {
+			r.err = fmt.Errorf("parsing error, row has different number of columns than the table: %d vs %d", len(row)-1, len(r.columns))
+			return false
+		}
+	}
+	switch {
+	case row[0] == "":
+		switch parsingState {
```

Comment:
Having a separate library for decoding Flux CSV is definitely a good idea. 👍 
 I know some users of the current Go client use the `QueryTableResult` as a Flux parser to read their Flux exports.
 
 Used algorithms and patterns is always a subjective approach. Your `annotatedcsv.Reader`  parser looks indeed more simple by separating parsing of header and row. If the `Decode` is suitable for the reader (I don't see why not), it can replace the current `QueryResults`.
 
 What would be the next step here? 
 Should I wait for you to finish it or should I contribute to the package by adding test, fixes, etc to help you to finish it?
 Will it be later moved to the influxdata GitHub org?

Reply:
> What would be the next step here?
> Should I wait for you to finish it or should I contribute to the package by adding test, fixes, etc to help you to finish it?
> Will it be later moved to the influxdata GitHub org?

Would you be up for taking my code above and writing tests/fixes to bring it up to scratch? If not, np, I can do it, but I'm hoping it's understandable enough that I don't have to. It should be possible to re-purpose your existing tests, I think, which should save some time.

I think it's fine to have it as a sub-package within this module, as there are some influx-specific things in it (e.g. the `tag` type).

Reply:
_Conversation continued via emails_